### PR TITLE
VectorWidth with MFMA SourceSwap

### PR DESCRIPTION
**Pull request by TonyYHsieh**
Support VectorWidth with MFMA SourceSwap
- line of output stride can be VW(2) * thread(16) * bpe(8) in double case, which is perfect for buffer_store_dwordx4.
- rewrite shiftptr for 0 dim which support any size of GRVW & VW.

### CODE REVIEW

**Comment by aazz44ss** [comment]
Can you add some comments on VectorWidth implementation with mfma to Common.py?

File: Tensile/Components/LocalRead.py [hunk 1]
```diff
@@ -138,7 +138,8 @@ def __call__(self, writer, bufferIdx, iui, epsi, tP):
 
         numOffsets       = instruction.numOffsets
         blockWidth       = instruction.blockWidth
-        MIWaveGropuShape = [ kernel["MatrixInstM"] * kernel["MatrixInstBM"] * kernel["MIWaveGroup"][0], \
+        vectorWidth      = kernel["VectorWidth"] if kernel["SourceSwap"] else 1 # TODO: nonSwap VectorWidth
```
Comment (aazz44ss) [inline Tensile/Components/LocalRead.py:141]: why not rejecting vectorWidth without SourceSwap in SolutionStruction.py?

**Review by AlexBrownAMD — COMMENTED** [review]
Ran some benchmarks with this PR and performance seems to drop with this change. Does this PR's performance gains rely on being combined with the other changes you outlined? Or are there certain params I should try out to match previous performance?

File: Tensile/SolutionStructs.py [hunk 2]
```diff
@@ -1887,11 +1887,8 @@ def assignProblemIndependentDerivedParameters(state):
       state["MatrixInstBN"]        = state["MIBlock"][5]
 
       state["LocalSplitU"]         = 1
-      state["MIOutputVectorWidth"] = 1 if (state["MatrixInstM"] == 4 and state["ProblemType"]["DataType"].isDouble()) else 4
-      state["MIRegPerOut"]         = 2 if state["ProblemType"]["DataType"].isDouble() else 1
-
-      if state["ProblemType"]["DataType"].isDouble() and state["StoreVectorWidth"] != 1:
-          reject(state, "DGEMM MFMA currently requires StoreVectorWidth=1")
```
[the remark this answers is not in the document]
Comment (TonyYHsieh) [reply Tensile/SolutionStructs.py:1894]: update. 
Please help to try again

File: Tensile/Components/LocalRead.py [hunk 3]
```diff
@@ -138,7 +138,8 @@ def __call__(self, writer, bufferIdx, iui, epsi, tP):
...
+        MIWaveGropuShape = [ kernel["MatrixInstM"] * kernel["MatrixInstBM"] * kernel["MIWaveGroup"][0] * vectorWidth, \
```
[the remark this answers is not in the document]
Comment (TonyYHsieh) [reply Tensile/Components/LocalRead.py]: update

**Comment by TonyYHsieh** [comment]
> Ran some benchmarks with this PR and performance seems to drop with this change. Does this PR's performance gains rely on being combined with the other changes you outlined? Or are there certain params I should try out to match previous performance?

May you help to provide drop yams??

In our origin MFMA algorithm, continuous output of dgemm should be 1.
So it can't use buffer_load_dwordx2.

Origin develop branch set 4 to dgemm and do extra code in shiftptr.

I modify my change to let dgemm use buffer_load_dwordx2 by hack.
Which should sync to origin develop branch.

File: Tensile/Components/LocalRead.py [hunk 4]
```diff
@@ -148,11 +149,12 @@ def __call__(self, writer, bufferIdx, iui, epsi, tP):
             tileStride   = kernel["_DepthULds"] + LdsPad
             UnrollStride = 1
 
-        numVectorsPerTile = kernel["MIWaveTile"][tile01]
+        numReadPerTileVector = vectorWidth if (tc == "A") else 1
```
Comment (jichangjichang) [inline Tensile/Components/LocalRead.py:152]: should be "if (tile01 == 0) else 1"  ?

File: Tensile/Components/ShiftVectorComponents.py [hunk 5]
```diff
@@ -509,3 +510,225 @@ def __call__(self, writer, kernel, tP):
...
+            So we do shift back by below alorithm.
+            1. check if M_size % GlobalLoadVectorWidth != 0, return if == 0
+            2. decide which subgroup we need to shift, M_size(3) means 3/8 = group 0
+            3. decide which thread we need to shift, we have different groups of thread, (0-31) for first group, (32-63) for second group.
+            4. decide which shift block (subTile1) we want to shift. for ex [0-1], [1-2], we want to shift second subtile
+        """
+
+        kStr = ""
+
+        regPerElem      = 2 if kernel["ProblemType"]["DataType"].isDouble() else 1
```
Comment (jichangjichang) [inline Tensile/Components/ShiftVectorComponents.py:545]: kernel["ProblemType"]["ComputeDataType"].numRegisters() ?

File: Tensile/Components/ShiftVectorComponents.py [hunk 6]
```diff
@@ -509,3 +510,225 @@ def __call__(self, writer, kernel, tP):
...
+        glvw            = tP["glvw"]
+        numThreadInWave = writer.kernel["WavefrontSize"]
+        vectorWidth     = kernel["VectorWidth"] if kernel["SourceSwap"] else 1
+        numContOutput0  = 1 if kernel["SourceSwap"] else kernel["MIOutputVectorWidth"]
```
Comment (jichangjichang) [inline Tensile/Components/ShiftVectorComponents.py:549]: why need to check SourceSwap?
This function should for SourceSwap only

File: Tensile/Components/ShiftVectorComponents.py [hunk 7]
```diff
@@ -509,3 +510,225 @@ def __call__(self, writer, kernel, tP):
...
+
+        # gbReg: glvw block id
+        kStr += writer.comment("gbReg: glvw block id")
+        gbReg = writer.vgprPool.checkOut(1)
+        kStr += vectorStaticDivide(gbReg, wgMT, glvw, tmpVgpr, tmpSgpr)
+
+        # tgbReg: thread in glvw block
+        kStr += writer.comment("tgbReg: glvw block id")
+        tgbReg = writer.vgprPool.checkOut(1)
+        kStr += vectorStaticDivide(tgbReg, "Serial", threadInterval, tmpVgpr, tmpSgpr)
```
Comment (jichangjichang) [inline Tensile/Components/ShiftVectorComponents.py:632]: should we leave commets

File: Tensile/SolutionStructs.py [hunk 8]
```diff
@@ -1887,11 +1887,11 @@ def assignProblemIndependentDerivedParameters(state):
       state["MatrixInstBN"]        = state["MIBlock"][5]
 
       state["LocalSplitU"]         = 1
-      state["MIOutputVectorWidth"] = 1 if (state["MatrixInstM"] == 4 and state["ProblemType"]["DataType"].isDouble()) else 4
-      state["MIRegPerOut"]         = 2 if state["ProblemType"]["DataType"].isDouble() else 1
+      state["MIOutputVectorWidth"] = 4 if not state["ProblemType"]["DataType"].isDouble() else 1
+      state["MIRegPerOut"]         = 1 if not state["ProblemType"]["DataType"].isDouble() else 2
```
Comment (jichangjichang) [inline Tensile/SolutionStructs.py:1891]: state["ProblemType"]["ComputeDataType"].numRegisters() ?

**Review by jichangjichang — COMMENTED** [review]
please check

File: Tensile/Components/LraTileAssignment.py [hunk 9]
```diff
@@ -136,6 +137,8 @@ def __call__(self, writer, kernel, tP):
             "2. block offset: bnOffset = bnIdx * strideBlock(%u)" % strideBlock)
         kStr += inst("_v_add_u32", vgpr(tReg), vgpr(wReg), vgpr(tReg), \
             "3. add N and block offset: bnOffset = block and N offset")
+        kStr += staticMultiply(vgpr(tReg), vgpr(tReg), vectorWidth, sgpr(tmpSgpr), \
+            "3. apply VectorWidth: bnOffset = bnOffset * vw(%u)" % vectorWidth)
```
[the remark this answers is not in the document]
Comment (TonyYHsieh) [reply Tensile/Components/LraTileAssignment.py:141]: 3

File: Tensile/Components/LocalRead.py (hunk 4 above)
Comment (TonyYHsieh) [reply Tensile/Components/LocalRead.py:152]: OK

File: Tensile/Components/ShiftVectorComponents.py (hunk 5 above)
Comment (TonyYHsieh) [reply Tensile/Components/ShiftVectorComponents.py:545]: no,   d gemm equal to z, c gemm equal to s.

File: Tensile/Components/ShiftVectorComponents.py (hunk 6 above)
Comment (TonyYHsieh) [reply Tensile/Components/ShiftVectorComponents.py:549]: this function is able to handle for both SourceSwap/non-SourceSwap.
I will update more code later

File: Tensile/Components/ShiftVectorComponents.py (hunk 7 above)
Comment (TonyYHsieh) [reply Tensile/Components/ShiftVectorComponents.py:632]: hard to comments

File: Tensile/SolutionStructs.py (hunk 8 above)
Comment (TonyYHsieh) [reply Tensile/SolutionStructs.py:1891]: no, as mention before.

**Review by aazz44ss — APPROVED** [review]
LGTM

**Comment by AlexBrownAMD** [comment]
> > Ran some benchmarks with this PR and performance seems to drop with this change. Does this PR's performance gains rely on being combined with the other changes you outlined? Or are there certain params I should try out to match previous performance?
> 
> May you help to provide drop yams??
> 
> In our origin MFMA algorithm, continuous output of dgemm should be 1.
> So it can't use buffer_load_dwordx2.
> 
> Origin develop branch set 4 to dgemm and do extra code in shiftptr.
> 
> I modify my change to let dgemm use buffer_load_dwordx2 by hack.
> Which should sync to origin develop branch.

Still seeing slowdowns with latest changes, benchmarked a wide range of yaml parameters. Can you merge this and other dgemm performance changes (like PR 1351) into the branch DgemmExp se we can test them together? If you could also provide a yaml file you've used to get best results that'd help. Thanks!

**Review by AlexBrownAMD — APPROVED** [review]
LGTM

**Pull request merged**